pkg/server/sotw/v3/server.go

```diff
@@ -176,6 +184,16 @@ func (s *server) process(stream Stream, reqCh <-chan *discovery.DiscoveryRequest
 		// increment nonce
 		streamNonce = streamNonce + 1
 		out.Nonce = strconv.FormatInt(streamNonce, 10)
+
+		lastResponse := latestDiscoveryResponse{
+			nonce:     out.Nonce,
+			resources: make(map[string]struct{}),
+		}
+		for _, r := range resp.(*cache.RawResponse).Resources {
```

Comment:
i think here we should use the ResourceNames from original request rather than the response resources list. because there is chance the response list contain more names than the request list (which although not make sense but still valid logic). in that case we still want resent the response if the new request adding more names already sent before.

Reply:
I was thinking about it as well. But I believe the implemented logic has more sense (less confusing) since it implies the tracking of resource names which are sent back to the caller. Java control plane implementation implements exactly same logic as I did in PR.

Reply:
it does not fully follow the protocol, which means the java implementation also buggy. this is what the protocol defined:

> When the client sends a new request that changes the set of resources being requested, the server must resend any newly requested resources, even if it previously sent those resources without having been asked for them and the resources have not changed since that time.

Reply:
I see. It is easy to change the logic but it will be hard to test in unit-test since the implementation sends only requested resources. Shall I change it ? Any other thoughts

Reply:
I think you can get the original request by RawResponse.Request object, and you only need compare the names.

Reply:
Response has GetRequest method that gives the original request. As I said it is easy to change

Reply:
Changed in https://github.com/envoyproxy/go-control-plane/pull/490/commits/1df2abd6aebfce09974db904e0ba47e6843f7c89